### ISSUE TITLE
The problem of the intersection of person ids betweem trainval and gallery

### ISSUE THREAD

**Issue by Burrocode** [huanghoujing/person-reid-triplet-loss-baseline#6]
Thank you for sharing the code!

when I write the codes as follow ,
```      
if osp.exists(train_test_partition_file):
            train_test_partition = load_pickle(train_test_partition_file)
        else:
            raise RuntimeError("Train/test partition file should be provided.'") 
trainval_pids=set()
query_pids = set()
gallery_pids = set()
for im_type in ['detected','labeled']:
            trainval_im_names = train_test_partition[im_type]['train_im_names']
            query_im_names = train_test_partition[im_type]['query_im_names']
            gallery_im_names = train_test_partition[im_type]['gallery_im_names']
            
            trainval_pids.update(set([parse_im_name(n,'id') for n in trainval_im_names]))
            
            query_pids.update(set([parse_im_name(n,'id') for n in query_im_names]))
            gallery_pids.update(set([parse_im_name(n,'id') for n in gallery_im_names]))
           
        
print (trainval_pids & gallery_pids)
assert query_pids <= gallery_pids
assert trainval_pids.isdisjoint(gallery_pids)
```

It causes "AssertionError".
I find that there is an intersection between trainval and gallery. (set([1201, 1389]))
The training/testing partition file is “re_ranking_train_test_split.pkl” downloaded from the given link. 
Looking forward for your reply, thank you.

**Comment by huanghoujing** [comment]
Thank you a lot for pointing out the problem. I will look into that!

**Comment by huanghoujing** [comment]
1. In your code, `trainval_pids` would be updated by both `detected` and `labeled` parts of the dataset, similarly for `query_pids` and `gallery_pids`. When you check the `detected` and `labeled` parts independently, as in the modified code below, you would find there is no problem.

```python
if osp.exists(train_test_partition_file):
  train_test_partition = load_pickle(train_test_partition_file)
else:
  raise RuntimeError("Train/test partition file should be provided.'")

for im_type in ['detected', 'labeled']:
  trainval_pids = set()
  query_pids = set()
  gallery_pids = set()

  trainval_im_names = train_test_partition[im_type]['train_im_names']
  query_im_names = train_test_partition[im_type]['query_im_names']
  gallery_im_names = train_test_partition[im_type]['gallery_im_names']

  trainval_pids.update(set([parse_im_name(n, 'id') for n in trainval_im_names]))
  query_pids.update(set([parse_im_name(n, 'id') for n in query_im_names]))
  gallery_pids.update(set([parse_im_name(n, 'id') for n in gallery_im_names]))

  print(trainval_pids & gallery_pids)
  assert query_pids <= gallery_pids
  assert trainval_pids.isdisjoint(gallery_pids)
```

2. Your code indicates the inconsistency between the `detected` partitions and `labeled` partitions. This originates from the [official new train/test protocol](https://github.com/zhunzhong07/person-re-ranking). I provide [a script](https://github.com/huanghoujing/person-re-ranking/blob/32d8e55be726c2c9944a3e759333f787312cefc6/CUHK03-Map-New-Protocol-To-My-Saved-Image-Names/check_new_protocol_official_files.py) to verify this. Where you will find that
- `detected` train ids share 766 ids with `labeled` train ids
- `detected` test ids share 699 ids with `labeled` test ids
- id `(4, 30)` (id `1201` in my project) is in train ids for `detected`, but in test ids for `labeled`
- id `(2, 359)` (id `1389` in my project) is in test ids for `detected`, but in train ids for `labeled`

3. To make the generation of my `re_ranking_train_test_split.pkl` more clear, and to make future problems more trackable, I provide details of the generation process in [this repository](https://github.com/huanghoujing/person-re-ranking/tree/32d8e55be726c2c9944a3e759333f787312cefc6/CUHK03-Map-New-Protocol-To-My-Saved-Image-Names).
